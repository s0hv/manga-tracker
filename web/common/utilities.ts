import { Group } from '@/webUtils/utilities';

type GetKey<T> = (value: T) => string;
interface GroupByOptions<B extends boolean = boolean, C extends boolean = boolean> {
  keepOrder: B
  returnAsDict?: C
}

type GroupBy = {
  <T, B extends true = true, C extends false = false>(arr: T[], getKeyOrKey: keyof T | GetKey<T>, options?: GroupByOptions<B, C>): Group<T>[]
  <T, B extends false = false, C extends false = false>(arr: T[], getKeyOrKey: keyof T | GetKey<T>, options?: GroupByOptions<B, C>): T[][]
  <T, B extends false = false, C extends true = true>(arr: T[], getKeyOrKey: keyof T | GetKey<T>, options?: GroupByOptions<B, C>): {[key: string]: T[]}
}

/**
 *
 * @param arr
 * @param getKeyOrKey Either an object property name or a function that returns the group key
 * @param options
 * @param options.keepOrder If true the order of the array will be kept the same
 * meaning the same key can be grouped multiple times
 * @param options.returnAsDict If true returns the grouped values as a dictionary. Ignored if keepOrder is true.
 */
export const groupBy: GroupBy = <T, >(
  arr: T[],
  getKeyOrKey: keyof T | GetKey<T>,
  { keepOrder = true, returnAsDict = false } = {} as GroupByOptions<true, false>
  // By adding "| T" to the return type somehow fixes overload errors WTF???
  // GJ typescript
): Group<T>[] | T[][] | T | {[key: string]: T[]} => {
  if (!Array.isArray(arr)) {
    throw new TypeError('Input must be an array');
  }

  if (arr.length === 0) {
    return returnAsDict ? {} : [];
  }

  let getKey: GetKey<T>;
  if (typeof getKeyOrKey === 'string') {
    const _key = getKeyOrKey;
    getKey = (o) => (o as any)[_key];
  } else {
    getKey = getKeyOrKey as any;
  }

  if (keepOrder) {
    const groups: Group<T>[] = [];
    let currentGroup: Group<T> = {
      group: '',
      arr: [],
    };

    let currentGroupKey: string;
    arr.forEach((o) => {
      const key = getKey(o);
      if (key !== currentGroupKey) {
        if (currentGroup.arr.length > 0) {
          groups.push(currentGroup);
        }
        currentGroup = {
          group: '',
          arr: [],
        };
        currentGroup.group = key;
        currentGroupKey = key;
      }

      currentGroup.arr.push(o);
    });

    if (currentGroup.arr.length > 0) {
      currentGroup.group = currentGroupKey!;
      groups.push(currentGroup);
    }

    return groups;
  }

  // Set is iterated in insertion order https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set#description
  const order: Set<string> = new Set();
  const group: Record<string, T[]> = {};

  arr.forEach((o) => {
    const key = getKey(o);
    // If group made add to that
    if (order.has(key)) {
      group[key].push(o);
    } else {
      // No group yet. Create new group and add group key to order
      group[key] = [o];
      order.add(key);
    }
  });

  if (returnAsDict) {
    return group;
  }

  const retVal: T[][] = [];
  order.forEach((groupKey) => retVal.push(group[groupKey]));
  return retVal;
};
