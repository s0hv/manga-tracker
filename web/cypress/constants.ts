export const nextAuthSessionCookie = 'next-auth.session-token';
