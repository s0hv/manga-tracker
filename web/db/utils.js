import {
  NUMERIC_VALUE_OUT_OF_RANGE,
  INVALID_TEXT_REPRESENTATION,
  UNIQUE_VIOLATION,
  IN_FAILED_SQL_TRANSACTION,
  FOREIGN_KEY_VIOLATION,
  NOT_NULL_VIOLATION,
} from 'pg-error-constants';

import { pgp } from '.';
import { NoColumnsError } from './errors.js';
import { dbLogger } from '../utils/logging.js';
import { StatusError } from '../utils/errors.js';

/**
 * Generate update statement from an object while filtering out undefined values.
 * The generated update statement lacks a WHERE clause and has the values embedded in the query.
 * @param {Object} o Input object
 * @param {String} tableName Name of the table to be updated
 * @returns {String} The update statement without a whereclause
 */
export const generateUpdate = (o, tableName) => {
  const obj = { ...o };
  Object.keys(obj).forEach(key => obj[key] === undefined && delete obj[key]);
  try {
    return pgp.helpers.update(obj, null, tableName);
  } catch (e) {
    throw new NoColumnsError('No valid columns given');
  }
};

export function handleError(err, res, msgOverrides = {}) {
  if (typeof err?.getErrors === 'function') {
    err = err.getErrors().filter(e => e?.code !== IN_FAILED_SQL_TRANSACTION)[0] || err;
  }

  if (err instanceof StatusError) {
    res.status(err.status).json({ error: err.message });
    return;
  }

  if (err instanceof NoColumnsError) {
    res.status(400).json({ error: err.message || 'No valid columns given' });
    return;
  }

  const msg = msgOverrides[err.code];
  if (err.code === INVALID_TEXT_REPRESENTATION) {
    dbLogger.debug(err.message);
    res.status(400).json({ error: msg || 'Invalid data type given' });
  } else if (err.code === NUMERIC_VALUE_OUT_OF_RANGE) {
    res.status(400).json({ error: msg || 'Number value out of range' });
  } else if (err.code === UNIQUE_VIOLATION) {
    res.status(422).json({ error: msg || 'Resource already exists' });
  } else if (err.code === FOREIGN_KEY_VIOLATION) {
    res.status(404).json({ error: msg || 'Foreign key violation' });
  } else if (err.code === NOT_NULL_VIOLATION) {
    res.status(400).json({ error: msg || 'Not null value was null' });
  } else {
    dbLogger.error(err, 'Unknown database error');
    res.status(500).json({ error: msg || 'Internal server error' });
  }
}
