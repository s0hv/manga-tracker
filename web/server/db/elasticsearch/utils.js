export const handleElasticError = (err, res) => {
  const error = 'Internal server error';
  console.error(err);
  res.status(500).json({ error });
};

export const extractFields = (result, fields, idPrefix, customFieldFormatter) => {
  idPrefix = idPrefix ? `${idPrefix}_` : '';
  return result.body.hits.hits.map(hit => {
    const o = {
      [`${idPrefix}id`]: Number(hit._id),
      score: hit._score,
    };
    if (fields) {
      fields.forEach(field => {
        o[field] = (hit.fields[field] || [])[0];
      });

      if (customFieldFormatter) {
        const formatted = customFieldFormatter(hit.fields);
        Object.keys(formatted).forEach(k => { o[k] = formatted[k] });
      }
    }

    return o;
  });
};
