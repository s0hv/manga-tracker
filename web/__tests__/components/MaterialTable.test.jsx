import React from 'react';
import fetchMock from 'fetch-mock';
import { render, screen, act, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Checkbox } from '@material-ui/core';
import {
  defaultOnSaveRow,
  EditableCheckbox,
  EditableDateTimePicker,
  MaterialTable,
} from '../../src/components/MaterialTable';
import { mockUTCDates, withRoot } from '../utils';
import { defaultDateFormatRegex } from '../constants';
import { defaultDateFormat } from '../../src/utils/utilities';

fetchMock.config.overwriteRoutes = true;

const columns = [
  { Header: 'ID', accessor: 'id', canEdit: false },
  { Header: 'Editable string', accessor: 'editableString' },
  {
    Header: 'Editable time',
    accessor: 'editableTime',
    sortType: 'datetime',
    Cell: ({ row }) => defaultDateFormat(row.values.editableTime),
    EditCell: ({ row, state, cell }) => (
      <EditableDateTimePicker
        clearable={1}
        variant='inline'
        ampm={false}
        value={row.values.editableTime}
        row={row}
        state={state}
        cell={cell}
      />
    ),
  },
  {
    Header: 'Editable checkbox',
    accessor: 'editableCheckbox',
    sortType: 'basic',
    Cell: ({ row }) => <Checkbox checked={row.values.editableCheckbox} disabled />,
    EditCell: ({ row, state, cell }) => (
      <EditableCheckbox
        checked={row.values.editableCheckbox}
        row={row}
        state={state}
        cell={cell}
      />
    ),
  },
];

function createRow(id, editableString, editableTime, editableCheckbox) {
  return {
    id,
    editableString,
    editableTime,
    editableCheckbox,
  };
}

const data = [
  createRow('unique_id1', 'test string', new Date('2020-07-15T15:51:17.885Z'), false),
  createRow('unique_id2', 'test string 2', new Date('2019-07-15T15:51:17.885Z'), true),
  createRow('unique_id3', 'test string 3', new Date('2020-09-15T15:51:17.885Z'), false),
];

function createWrapper(props) {
  render(
    withRoot(
      <MaterialTable {...props} />
    )
  );
}

describe('It should render correctly', () => {
  mockUTCDates();

  const expectNoLoadingElements = () => {
    expect(screen.queryByLabelText(/loading icon/i)).not.toBeInTheDocument();
  };

  const expectNoEditElements = () => {
    expect(screen.queryByRole('button', { name: /edit row/i })).not.toBeInTheDocument();
    expect(screen.queryByRole('button', { name: /delete row/i })).not.toBeInTheDocument();
  };

  const expectEditElementsExist = () => {
    expect(screen.getAllByRole('button', { name: /edit row/i })).toBeTruthy();
    expect(screen.getAllByRole('button', { name: /delete row/i })).toBeTruthy();
    expect(screen.getAllByRole('button', { name: /add item/i })).toBeTruthy();
  };

  const expectHeadersExist = (headers = columns) => {
    headers.forEach(col => {
      expect(screen.getByRole('columnheader', { name: col.Header })).toBeInTheDocument();
    });
  };

  test('without data', () => {
    render(
      <MaterialTable
        columns={columns}
        data={[]}
      />
    );

    // Only header row should exist
    expect(screen.getAllByRole('row')).toHaveLength(1);

    expect(screen.queryByRole('button')).not.toBeInTheDocument();

    expectNoLoadingElements();
    expectNoEditElements();
    expectHeadersExist();
  });

  test('without data (null)', () => {
    expect(() => render(
      <MaterialTable
        columns={columns}
        data={null}
      />
    )).toThrow(TypeError);
  });

  test('with data', () => {
    render(
      <MaterialTable
        columns={columns}
        data={data}
        editable
        deletable
        creatable
        sortable
        pagination
        CreateDialog={() => null}
      />
    );

    // Make sure sort buttons exist
    columns.forEach(col => {
      expect(screen.getByRole('button', { name: col.Header }));
    });

    // Data rows + header row
    expect(screen.getAllByRole('row')).toHaveLength(data.length + 1);

    expectNoLoadingElements();

    expectEditElementsExist();

    screen.getAllByRole('row')
      .slice(1)
      .forEach((rowElem, idx) => {
        const row = within(rowElem);
        const values = data[idx];

        expect(row.getByRole('cell', { name: values.id })).toBeInTheDocument();
        expect(row.getByRole('cell', { name: values.editableString })).toBeInTheDocument();
        expect(row.getByRole('cell', { name: new RegExp(defaultDateFormatRegex, 'i') })).toBeInTheDocument();
        expect(row.getByRole('checkbox', { checked: values.editableCheckbox }));
      });

    // Test pagination element
    const pagination = within(screen.getByRole('navigation', { name: /table pagination/i }));

    expect(pagination.getByRole('button', { name: /first page/i })).toBeInTheDocument();
    expect(pagination.getByRole('button', { name: /previous page/i })).toBeInTheDocument();
    expect(pagination.getByRole('button', { name: /next page/i })).toBeInTheDocument();
    expect(pagination.getByRole('button', { name: /last page/i })).toBeInTheDocument();
  });

  test('when loading with existing rows', () => {
    render(
      <MaterialTable
        columns={columns}
        data={data}
        loading
        pagination
      />
    );

    expect(screen.getByRole('progressbar', { name: /loading icon/i }));
    // No skeletons should be visible
    expect(screen.getAllByRole('row', { hidden: true })).toHaveLength(data.length + 1);
  });

  test('when loading without existing rows', () => {
    render(
      <MaterialTable
        columns={columns}
        data={[]}
        loading
        pagination
      />
    );

    expect(screen.getByRole('progressbar', { name: /loading icon/i }));
    expect(screen.getAllByRole('row')).toHaveLength(1);

    // Skeletons should be rendered
    expect(screen.getAllByRole('row', { hidden: true }).length).toBeGreaterThan(1);
  });
});

describe('Should handle editing', () => {
  test('should call save when save clicked', async () => {
    const onSave = jest.fn();

    createWrapper({
      columns,
      data,
      onSaveRow: onSave,
      editable: true,
    });

    const row = within(screen.getAllByRole('row')[1]);
    expect(row).toBeDefined();

    act(() => {
      userEvent.click(
        row.getByRole('button', { name: /edit row/i })
      );
    });

    act(() => {
      userEvent.click(
        row.getByRole('button', { name: /save row/i })
      );
    });

    expect(onSave).toHaveBeenCalledTimes(1);
    expect(screen.queryByRole('button', { name: /cancel edit/i })).not.toBeInTheDocument();
  });

  it('should not call save when cancel clicked', async () => {
    const onSave = jest.fn();

    createWrapper({
      columns,
      data,
      onSaveRow: onSave,
      editable: true,
    });

    const row = within(screen.getAllByRole('row')[1]);
    expect(row).toBeDefined();

    act(() => {
      userEvent.click(
        row.getByRole('button', { name: /edit row/i })
      );
    });

    act(() => {
      userEvent.click(
        row.getByRole('button', { name: /cancel edit/i })
      );
    });

    expect(onSave).not.toHaveBeenCalled();
    expect(screen.queryByRole('button', { name: /save row/i })).not.toBeInTheDocument();
    expect(screen.queryByRole('button', { name: /cancel edit/i })).not.toBeInTheDocument();
  });

  it('should not allow editing non editable columns', async () => {
    const onSave = jest.fn();

    createWrapper({
      columns,
      data,
      onSaveRow: onSave,
      editable: true,
    });

    expect(columns[0].canEdit).toBeFalse();

    const row = within(screen.getAllByRole('row')[1]);
    expect(row).toBeDefined();

    act(() => {
      userEvent.click(
        row.getByRole('button', { name: /edit row/i })
      );
    });

    const cell = within(row.getByText(data[0].id).closest('td'));
    // Make sure that no input is present
    expect(cell.queryByRole('textbox')).not.toBeInTheDocument();
  });

  it('should allow editing string columns', async () => {
    const onSave = jest.fn().mockImplementation(defaultOnSaveRow);

    createWrapper({
      columns,
      data,
      onSaveRow: onSave,
      editable: true,
    });

    const row = within(screen.getAllByRole('row')[1]);
    expect(row).toBeDefined();

    act(() => {
      userEvent.click(
        row.getByRole('button', { name: /edit row/i })
      );
    });

    // Find input and change its value
    const input = row.getByLabelText(/editable string input/i);

    const newVal = 'value changed';

    await act(async () => {
      userEvent.clear(input);
      await userEvent.type(input, newVal, { delay: 1 });
    });

    act(() => {
      userEvent.click(
        row.getByRole('button', { name: /save row/i })
      );
    });

    expect(onSave).toHaveBeenCalledTimes(1);
    expect(onSave).toHaveBeenCalledWith(expect.anything(), { editableString: newVal }, expect.anything());

    expect(row.getByText(newVal)).toBeInTheDocument();
  });

  it('should allow editing checkbox', async () => {
    const onSave = jest.fn().mockImplementation(defaultOnSaveRow);

    createWrapper({
      columns,
      data,
      onSaveRow: onSave,
      editable: true,
    });

    const row = within(screen.getAllByRole('row')[1]);
    expect(row).toBeDefined();

    act(() => {
      userEvent.click(
        row.getByRole('button', { name: /edit row/i })
      );
    });

    // Find checkbox and click it to change it's value
    const checkbox = row.getByRole('checkbox', { checked: data[0].editableCheckbox });
    userEvent.click(checkbox);

    act(() => {
      userEvent.click(
        row.getByRole('button', { name: /save row/i })
      );
    });

    expect(onSave).toHaveBeenCalledTimes(1);
    expect(onSave).toHaveBeenCalledWith(expect.anything(), { editableCheckbox: !data[0].editableCheckbox }, expect.anything());
  });
});
