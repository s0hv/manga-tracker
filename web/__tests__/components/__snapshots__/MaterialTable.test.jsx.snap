// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It should render correctly with data 1`] = `
Array [
  <div>
    <div
      className="MuiToolbar-root MuiToolbar-regular makeStyles-root-5 MuiToolbar-gutters"
    >
      <h6
        className="MuiTypography-root makeStyles-title-6 MuiTypography-h6"
      />
    </div>
    <table
      className="MuiTable-root"
      role="table"
    >
      <colgroup>
        <col
          key="id"
        />
        <col
          key="editableString"
        />
        <col
          key="editableTime"
        />
        <col
          key="editableCheckbox"
        />
      </colgroup>
      <thead
        className="MuiTableHead-root makeStyles-tableHead-2"
        role={null}
      >
        <tr
          className="MuiTableRow-root MuiTableRow-head"
          role="row"
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head"
            colSpan={1}
            role="columnheader"
            scope="col"
            style={
              Object {
                "cursor": undefined,
              }
            }
          />
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head"
            colSpan={1}
            onClick={[Function]}
            role="columnheader"
            scope="col"
            style={
              Object {
                "cursor": "pointer",
              }
            }
            title="Toggle SortBy"
          >
            <span
              aria-disabled={false}
              className="MuiButtonBase-root MuiTableSortLabel-root"
              onBlur={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              role="button"
              tabIndex={0}
            >
              ID
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                />
              </svg>
            </span>
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head"
            colSpan={1}
            onClick={[Function]}
            role="columnheader"
            scope="col"
            style={
              Object {
                "cursor": "pointer",
              }
            }
            title="Toggle SortBy"
          >
            <span
              aria-disabled={false}
              className="MuiButtonBase-root MuiTableSortLabel-root"
              onBlur={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              role="button"
              tabIndex={0}
            >
              Editable string
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                />
              </svg>
            </span>
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head"
            colSpan={1}
            onClick={[Function]}
            role="columnheader"
            scope="col"
            style={
              Object {
                "cursor": "pointer",
              }
            }
            title="Toggle SortBy"
          >
            <span
              aria-disabled={false}
              className="MuiButtonBase-root MuiTableSortLabel-root"
              onBlur={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              role="button"
              tabIndex={0}
            >
              Editable time
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                />
              </svg>
            </span>
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head"
            colSpan={1}
            onClick={[Function]}
            role="columnheader"
            scope="col"
            style={
              Object {
                "cursor": "pointer",
              }
            }
            title="Toggle SortBy"
          >
            <span
              aria-disabled={false}
              className="MuiButtonBase-root MuiTableSortLabel-root"
              onBlur={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              role="button"
              tabIndex={0}
            >
              Editable checkbox
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                />
              </svg>
            </span>
          </th>
        </tr>
      </thead>
      <tbody
        className="MuiTableBody-root"
        role="rowgroup"
      >
        <tr
          className="MuiTableRow-root"
          role="row"
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingCheckbox"
            role="cell"
          >
            <div
              className="makeStyles-editCell-1"
            >
              <button
                aria-label="edit row"
                className="MuiButtonBase-root MuiIconButton-root"
                disabled={false}
                name="edit"
                onBlur={[Function]}
                onClick={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={0}
                type="button"
              >
                <span
                  className="MuiIconButton-label"
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                    />
                  </svg>
                </span>
                <span
                  className="MuiTouchRipple-root"
                />
              </button>
            </div>
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            role="cell"
          >
            unique_id1
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            role="cell"
          >
            test string
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            role="cell"
          >
            Jul 15th, 15:51
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            role="cell"
          >
            <span
              aria-disabled={true}
              className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-7 MuiCheckbox-root MuiCheckbox-colorSecondary PrivateSwitchBase-disabled-9 Mui-disabled MuiIconButton-colorSecondary Mui-disabled Mui-disabled"
              onBlur={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={-1}
            >
              <span
                className="MuiIconButton-label"
              >
                <input
                  checked={false}
                  className="PrivateSwitchBase-input-10"
                  data-indeterminate={false}
                  disabled={true}
                  onChange={[Function]}
                  type="checkbox"
                />
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                  />
                </svg>
              </span>
            </span>
          </td>
        </tr>
        <tr
          className="MuiTableRow-root"
          role="row"
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingCheckbox"
            role="cell"
          >
            <div
              className="makeStyles-editCell-1"
            >
              <button
                aria-label="edit row"
                className="MuiButtonBase-root MuiIconButton-root"
                disabled={false}
                name="edit"
                onBlur={[Function]}
                onClick={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={0}
                type="button"
              >
                <span
                  className="MuiIconButton-label"
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                    />
                  </svg>
                </span>
                <span
                  className="MuiTouchRipple-root"
                />
              </button>
            </div>
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            role="cell"
          >
            unique_id2
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            role="cell"
          >
            test string 2
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            role="cell"
          >
            Jul 15th, 15:51
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            role="cell"
          >
            <span
              aria-disabled={true}
              className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-7 MuiCheckbox-root MuiCheckbox-colorSecondary PrivateSwitchBase-checked-8 Mui-checked PrivateSwitchBase-disabled-9 Mui-disabled MuiIconButton-colorSecondary Mui-disabled Mui-disabled"
              onBlur={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={-1}
            >
              <span
                className="MuiIconButton-label"
              >
                <input
                  checked={true}
                  className="PrivateSwitchBase-input-10"
                  data-indeterminate={false}
                  disabled={true}
                  onChange={[Function]}
                  type="checkbox"
                />
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                  />
                </svg>
              </span>
            </span>
          </td>
        </tr>
        <tr
          className="MuiTableRow-root"
          role="row"
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingCheckbox"
            role="cell"
          >
            <div
              className="makeStyles-editCell-1"
            >
              <button
                aria-label="edit row"
                className="MuiButtonBase-root MuiIconButton-root"
                disabled={false}
                name="edit"
                onBlur={[Function]}
                onClick={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={0}
                type="button"
              >
                <span
                  className="MuiIconButton-label"
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                    />
                  </svg>
                </span>
                <span
                  className="MuiTouchRipple-root"
                />
              </button>
            </div>
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            role="cell"
          >
            unique_id3
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            role="cell"
          >
            test string 3
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            role="cell"
          >
            Sep 15th, 15:51
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            role="cell"
          >
            <span
              aria-disabled={true}
              className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-7 MuiCheckbox-root MuiCheckbox-colorSecondary PrivateSwitchBase-disabled-9 Mui-disabled MuiIconButton-colorSecondary Mui-disabled Mui-disabled"
              onBlur={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={-1}
            >
              <span
                className="MuiIconButton-label"
              >
                <input
                  checked={false}
                  className="PrivateSwitchBase-input-10"
                  data-indeterminate={false}
                  disabled={true}
                  onChange={[Function]}
                  type="checkbox"
                />
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                  />
                </svg>
              </span>
            </span>
          </td>
        </tr>
      </tbody>
    </table>
  </div>,
  "",
]
`;

exports[`It should render correctly without data 1`] = `
Array [
  <div>
    <div
      className="MuiToolbar-root MuiToolbar-regular makeStyles-root-5 MuiToolbar-gutters"
    >
      <h6
        className="MuiTypography-root makeStyles-title-6 MuiTypography-h6"
      />
    </div>
    <table
      className="MuiTable-root"
      role="table"
    >
      <colgroup>
        <col
          key="id"
        />
        <col
          key="editableString"
        />
        <col
          key="editableTime"
        />
        <col
          key="editableCheckbox"
        />
      </colgroup>
      <thead
        className="MuiTableHead-root makeStyles-tableHead-2"
        role={null}
      >
        <tr
          className="MuiTableRow-root MuiTableRow-head"
          role="row"
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head"
            colSpan={1}
            role="columnheader"
            scope="col"
            style={
              Object {
                "cursor": undefined,
              }
            }
          >
            ID
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head"
            colSpan={1}
            role="columnheader"
            scope="col"
            style={
              Object {
                "cursor": undefined,
              }
            }
          >
            Editable string
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head"
            colSpan={1}
            role="columnheader"
            scope="col"
            style={
              Object {
                "cursor": undefined,
              }
            }
          >
            Editable time
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head"
            colSpan={1}
            role="columnheader"
            scope="col"
            style={
              Object {
                "cursor": undefined,
              }
            }
          >
            Editable checkbox
          </th>
        </tr>
      </thead>
      <tbody
        className="MuiTableBody-root"
        role="rowgroup"
      />
    </table>
  </div>,
  "",
]
`;
